server/channels/jobs/batch_migration_worker.go

```diff
@@ -0,0 +1,234 @@
+
+// DoJob executes the job picked up through the job channel.
+//
+// Note that this is a lot of distracting machinery here to claim the job, then double check the
+// status, and keep the status up to date in line with job infrastrcuture semantics. Unless an
+// error occurs, this worker should hold onto the job until its completed.
+func (worker *BatchMigrationWorker) DoJob(job *model.Job) {
+	logger := worker.logger.With(mlog.Any("job", job))
+	logger.Debug("Worker received a new candidate job.")
+	defer worker.jobServer.HandleJobPanic(logger, job)
+
+	if claimed, err := worker.jobServer.ClaimJob(job); err != nil {
+		logger.Warn("Worker experienced an error while trying to claim job", mlog.Err(err))
+		return
+	} else if !claimed {
+		return
+	}
+
+	c := request.EmptyContext(logger)
+	var appErr *model.AppError
+
+	// We get the job again because ClaimJob changes the job status.
+	job, appErr = worker.jobServer.GetJob(c, job.Id)
+	if appErr != nil {
+		worker.logger.Error("Worker: job execution error", mlog.Err(appErr))
+		worker.setJobError(logger, job, appErr)
+		return
+	}
+
+	if job.Data == nil {
+		job.Data = make(model.StringMap)
+	}
+
+	for {
+		select {
+		case <-worker.stop:
+			logger.Info("Worker: Migration has been canceled via Worker Stop. Setting the job back to pending.")
+			if err := worker.jobServer.SetJobPending(job); err != nil {
+				worker.logger.Error("Worker: Failed to mark job as pending", mlog.Err(err))
+			}
+			return
+		case <-time.After(worker.timeBetweenBatches):
+			// Ensure the cluster remains in sync, otherwise we restart the job to
+			// ensure a complete migration. Technically, the cluster could go out of
+			// sync briefly within a batch, but we accept that risk.
+			if !worker.checkIsClusterInSync() {
+				worker.logger.Warn("Worker: Resetting job")
+				worker.resetJob(logger, job)
+				return
+			}
+
+			worker.logger.Trace("Worker: processing batch", mlog.Any("data", job.Data))
+
+			nextData, done, err := worker.doMigrationBatch(job.Data, worker.store)
+			if err != nil {
+				worker.logger.Error("Worker: Failed to do migration batch. Exiting", mlog.Err(err))
+				worker.setJobError(logger, job, model.NewAppError("doMigrationsBatch", model.NoTranslation, nil, "", http.StatusInternalServerError).Wrap(err))
+				return
+			} else if done {
+				logger.Info("Worker: Job is complete")
+				worker.setJobSuccess(logger, job)
+				worker.markAsComplete()
+				return
+			}
+
+			job.Data = nextData
+
+			// Migrations currently don't support reporting meaningful progress.
+			worker.jobServer.SetJobProgress(job, 0)
+		}
+	}
+}
+
+// resetJob erases the data tracking the next batch to execute and returns the job status to
+// pending to allow the job infrastructure to requeue it.
+func (worker *BatchMigrationWorker) resetJob(logger mlog.LoggerIFace, job *model.Job) {
+	job.Data = nil
+	job.Progress = 0
+
+	if _, err := worker.store.Job().UpdateOptimistically(job, model.JobStatusInProgress); err != nil {
```

Comment:
Any reason we aren't directly setting the status to pending here? Seems like we are making yet another call just to set the status to pending.

Reply:
Good eye! I got wrapped in the subtleties of all the existing `setJob*` methods and missed that I could just set `job.Status = model.JobtStatusPending` directly above. Fixing!